perun-wui-registrar/src/main/java/cz/metacentrum/perun/wui/registrar/client/PerunRegistrarPresenter.java

```diff
@@ -137,103 +126,67 @@ public void onLoadingStart() {
-											if (event.getNativeKeyCode() == KeyCodes.KEY_ENTER) {
-												verify.click();
-											}
-										}
-									});
-
-									verify.addClickHandler(new ClickHandler() {
-										@Override
-										public void onClick(ClickEvent event) {
-											RegistrarManager.verifyCaptcha(captcha.getChallenge(), captcha.getResponse(), new JsonEvents() {
-												@Override
-												public void onFinished(JavaScriptObject result) {
-													verify.setProcessing(false);
-													BasicOverlayObject bt = result.cast();
-													if (bt.getBoolean()) {
-														group.setValidationState(ValidationState.SUCCESS);
-														modal.hide();
-													} else {
-														group.setValidationState(ValidationState.ERROR);
-													}
-												}
-
-												@Override
-												public void onError(PerunException error) {
-													verify.setProcessing(false);
-												}
-
-												@Override
-												public void onLoadingStart() {
-													verify.setProcessing(true);
-												}
-											});
-										}
-									});
-
-
-									body.add(captcha);
-
-									footer.add(form);
-
-									modal.add(header);
-									modal.add(body);
-									modal.add(footer);
-
-									modal.show();
-
-									Scheduler.get().scheduleDeferred(new Command() {
-										@Override
-										public void execute() {
-											response.setFocus(true);
-										}
-									});
-
-								}
-							}).setWindow(ScriptInjector.TOP_WINDOW).inject();
+					final Modal modal = new Modal();
+					modal.setRemoveOnHide(true);
+					modal.setDataBackdrop(ModalBackdrop.STATIC);
+
+					ModalHeader header = new ModalHeader();
+					header.setTitle(translation.pleaseVerifyCaptcha());
+					header.setClosable(false);
+					ModalBody body = new ModalBody();
+
+					final PerunLoader loader = new PerunLoader();
+					loader.onLoading();
+					loader.setVisible(false);
+					loader.getElement().getStyle().setMarginTop(20, Style.Unit.PX);
+
+					final ReCaptcha captcha = new ReCaptcha();
+					captcha.setSitekey(reCaptchaPublicKey);
+
+					final Alert alert = new Alert();
+					alert.setType(AlertType.DANGER);
+					alert.setVisible(false);
+
+					captcha.addResponseHandler(responseEvent ->
+											   RegistrarManager.verifyCaptcha(captcha.getResponse(), new JsonEvents() {
+						@Override
+						public void onFinished(JavaScriptObject result1) {
+							loader.setVisible(false);
+
+							BasicOverlayObject bt = result1.cast();
+							if (bt.getBoolean()) {
+								alert.setVisible(false);
+								modal.hide();
+							} else {
+								captcha.reset();
+								alert.setText(translation.captchaFailed());
+								alert.setVisible(true);
+							}
+						}
+
+						@Override
+						public void onError(PerunException error) {
+							loader.setVisible(false);
+							captcha.reset();
+							alert.setText(translation.serverDown());
```

Comment:
Can you please use our ``AlertErrorReporter`` which can interpret PerunException returned by RPC and allows user to report bug from GUI ? Thank you.

Reply:
Done.